File: grind-75/binary-tree/balanced-binary-tree.cc
#include <iostream>

/* -----------------------------------------------------------------------------
 * Link: https://leetcode.com/problems/balanced-binary-tree/
 * 
 * Solution Description:
 * We will traverse the tree using recursive DFS. Then in a bottom up manner we
 * will determine if each subtree is balanced by comparing the height of each
 * left and right subtree. This isBalanced property will bubble up to the root
 * to give us our final answer.
 *
 * Steps:
 * 1 - Define a recursive DFS helper function
 * a - When the function encounters null it should return (True, 0)
 * b - Otherwise the balanced boolean comes from calculating the absdiff between
 *     the heights of the left and right nodes if it is <= 1 then balanced is True
 * 2 - Call our helper function from our root and the returned boolean will be
 * 		 our final answer, we discard the final height
 *
 * Complexity Analysis:
 * Time: O(n) because in the worst case we traverse every node in the tree
 * Space: O(h) where h is the height of the tree, because in the worst case our
 * call stack grows to h for each level of the tree.
 *
 * -----------------------------------------------------------------------------
 */

struct SearchResult {
  bool balanced;
  int height;
};

SearchResult dfs(TreeNode* node) {
  if (node == nullptr) {
    return {true, 0};
  }
  const auto left{dfs(node->left)};
  const auto right{dfs(node->right)};
  return {
    left.balanced && right.balanced && std::abs(left.height - right.height) <= 1,
    std::max(left.height, right.height) + 1
  };
}

bool isBalanced(TreeNode* root) {
  return dfs(root).balanced;
}

File: grind-75/binary-tree/balanced-binary-tree.py
# ------------------------------------------------------------------------------
# See cc version for notes
# ------------------------------------------------------------------------------
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class SearchResult:
    def __init__(self, balanced: bool, height: int):
        self.balanced = balanced
        self.height = height

class Solution:
    def dfs(self, node: Optional[TreeNode]) -> SearchResult:
        if not node:
            return SearchResult(balanced=True, height=0)
        left, right = self.dfs(node=node.left), self.dfs(node=node.right)
        return SearchResult(
            balanced=left.balanced and right.balanced and abs(left.height - right.height) <= 1,
            height=max(left.height, right.height) + 1
        )

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root).balanced
